src/attribution/onINP.ts

```diff
@@ -274,6 +281,10 @@ const attributeINP = (metric: INPMetric): INPMetricWithAttribution => {
   );
 
   const nextPaintTime = Math.max.apply(Math, nextPaintTimeCandidates);
+  // If processingEnd has been capped to endTime then presentationDelay is 0
+  // Else use the nextPaintTime
+  const presentationDelay =
+    processingEnd == endTime ? 0 : Math.max(nextPaintTime - processingEnd, 0);
```

Comment:
I not sure this quite works, because I think it's possible that `endTime` is earlier than `nextPaintTime` (if the duration value is rounded down to the nearest 8ms), which would mean that the numbers wouldn't add up to the total INP duration.

I think it's safer to do all of the capping at the end rather than at various places through the logic. I also think it's potentially problematic (and confusing) to have both `endTime` and `nextPaintTime` and have those values possible be different things.

Reply:
Maybe the best solution here is to just do away with the logic that tries to be smart about figuring out more accurate `nextPaintTime` values and just use the values given.

Reply:
OK, I updated the code to do this. @tunetheweb let me know what you think.

Reply:
I like it. Much simpler.

My only comment is to wonder if we should apply the cap in `groupEntriesByRenderTime` instead? Because at the moment `getIntersectingLoAFs(group.startTime, group.processingEnd)` (as just one example) would include LoAFs after INP. For example:

- 10,000 ms: Interaction happens
- 10,016 ms: The interaction processing shows an `alert()`
- 10,030 ms: `alert()` is rendered and INP is capped.
- 11,000 ms: User dismisses the `alert()`

This produces a 1 second LoAF. Which technically is part of the INP interaction, but also technically doesn't matter that much (as INP ended before that), so is noise that doesn't need reporting (or maybe it's good to surface this?).

Capping `group.processingEnd` at the time it's set, would avoid that LoAF being included.

Also I'm not sure if in future people would forget to handle the fact that `group.processingEnd` can extend past INP time. Again capping `group.processingEnd` would mean anyone using `group` wouldn't have to consider that.

Reply:
The current code gets intersecting LoAFs based on the capped `processingEnd` rather than `group.processingEnd`, so I think it's already doing what you're suggesting (unless I'm misunderstanding you)?

Reply:
Oh so it does. The `cleanupEntries` code uses the `group.processingEnd` (which is not bad as meant it'll be kept cause it might be needed). My bad. But goes to show it's kinda confusing.

Reply:
Yeah, it is confusing...and I'm honestly not sure if it matters which one we choose because I think LoAF does report a long frame in this case, right? So that frame would get returned regardless of which processingEnd value was used?

Reply:
Agreed it is a LoAF. And I think it's right to be reported as a LoAF (though potentially not right to include that LoAF in INP attribution).

Anyway, I'm OK with the current implementation.
But I'm also OK with moving the capping to `groupEntriesByRenderTime` if that's less confusing (so you therefore wouldn't need the capping here as well).

WDYT is better?

Reply:
My main concern with moving the capping to `groupEntriesByRenderTime()` is that we don't currently have the "real" `renderTime` value in that function, because each entry has its own `startTime+duration` value. So to properly cap it there, we'd have to also store the `renderTime` for the group (which is okay, but would add additional code).

Reply:
OK let's leave as is then. Happy to merge.